tests/test_sft_trainer.py

```diff
@@ -654,6 +654,50 @@ def test_data_collator_completion_lm_with_multiple_text(self):
             result_text = tokenizer.decode(batch["input_ids"][i, last_pad_idx + 1 :])
             self.assertEqual(result_text, "I have not been masked correctly.")
 
+    def test_data_collator_completion_lm_without_padding(self):
+        os.environ["CUDA_VISIBLE_DEVICES"]="0"
+        model_id = "trl-internal-testing/tiny-random-LlamaForCausalLM"
+        torch_dtype = getattr(torch, "bfloat16", None)
+        model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch_dtype, attn_implementation="flash_attention_2")
+        tokenizer = AutoTokenizer.from_pretrained(model_id, use_fast=True)
+
+        formatted_dataset = lambda example: {
+            "output": f"### prompt:\n{example['prompt'].strip()}\n\n### completion:\n{example['completion'].strip()}{tokenizer.eos_token}"
+        }
```

Comment:
is dataset formatting required here, or can we drop it?

Reply:
dataset formatting is required because the `SFTTrainer` and `DataCollatorForCompletionOnlyLM` expect the dataset to have a specific format—a single text field that combines both the prompt and the completion in a way the model can understand. This function includes both the prompt and completion, ensuring the data collator can correctly identify where the completion starts using the `response_template`.